pkg/vault/test/foundry/E2eSwapRateProvider.t.sol

```diff
@@ -0,0 +1,142 @@
+            return rateProviderTokenB.getRate();
+        }
+    }
+
+    function testDoUndoExactInSwapRate__Fuzz(uint256 newRateTokenA, uint256 newRateTokenB) public {
+        _setPoolRates(newRateTokenA, newRateTokenB);
+
+        DoUndoLocals memory testLocals;
+        uint256 exactAmountIn = maxSwapAmountTokenA;
+
+        testDoUndoExactInBase(exactAmountIn, testLocals);
+    }
+
+    function testDoUndoExactInSwapRateComplete__Fuzz(
+        uint256 exactAmountIn,
+        uint256 poolSwapFeePercentage,
+        uint256 liquidityTokenA,
+        uint256 liquidityTokenB,
+        uint256 newDecimalsTokenA,
+        uint256 newDecimalsTokenB,
+        uint256 newRateTokenA,
+        uint256 newRateTokenB
+    ) public {
+        _setPoolRates(newRateTokenA, newRateTokenB);
+
+        DoUndoLocals memory testLocals;
+        testLocals.shouldTestDecimals = true;
+        testLocals.shouldTestLiquidity = true;
+        testLocals.shouldTestSwapAmount = true;
+        testLocals.shouldTestFee = true;
+        testLocals.liquidityTokenA = liquidityTokenA;
+        testLocals.liquidityTokenB = liquidityTokenB;
+        testLocals.newDecimalsTokenA = newDecimalsTokenA;
+        testLocals.newDecimalsTokenB = newDecimalsTokenB;
+        testLocals.poolSwapFeePercentage = poolSwapFeePercentage;
+
+        testDoUndoExactInBase(exactAmountIn, testLocals);
+    }
+
+    function testDoUndoExactOutSwapRate__Fuzz(uint256 newRateTokenA, uint256 newRateTokenB) public {
+        _setPoolRates(newRateTokenA, newRateTokenB);
+
+        DoUndoLocals memory testLocals;
+        uint256 exactAmountOut = maxSwapAmountTokenB;
+
+        testDoUndoExactOutBase(exactAmountOut, testLocals);
+    }
+
+    function testDoUndoExactOutSwapRateComplete__Fuzz(
+        uint256 exactAmountOut,
+        uint256 poolSwapFeePercentage,
+        uint256 liquidityTokenA,
+        uint256 liquidityTokenB,
+        uint256 newDecimalsTokenA,
+        uint256 newDecimalsTokenB,
+        uint256 newRateTokenA,
+        uint256 newRateTokenB
+    ) public {
+        _setPoolRates(newRateTokenA, newRateTokenB);
+
+        DoUndoLocals memory testLocals;
+        testLocals.shouldTestDecimals = true;
+        testLocals.shouldTestLiquidity = true;
+        testLocals.shouldTestSwapAmount = true;
+        testLocals.shouldTestFee = true;
+        testLocals.liquidityTokenA = liquidityTokenA;
+        testLocals.liquidityTokenB = liquidityTokenB;
+        testLocals.newDecimalsTokenA = newDecimalsTokenA;
+        testLocals.newDecimalsTokenB = newDecimalsTokenB;
+        testLocals.poolSwapFeePercentage = poolSwapFeePercentage;
+
+        testDoUndoExactOutBase(exactAmountOut, testLocals);
+    }
+
+    function _setPoolRates(uint256 newRateTokenA, uint256 newRateTokenB) private {
+        // Since the rate is an 18 decimals number, it ranges from 0.01 to 100.
+        newRateTokenA = bound(newRateTokenA, 1e16, 1e20);
+        newRateTokenB = bound(newRateTokenB, 1e16, 1e20);
```

Comment:
Does this break if we pump the max to 10k?

Reply:
No, I just changed the test to use up to 1 million without issues. But, I cannot go below 1e16, or else the math breaks.